pkg/controller/garbagecollector/rate_limiter_helper.go

```diff
@@ -29,31 +29,32 @@ import (
 // RegisteredRateLimiter records the registered RateLimters to avoid
 // duplication.
 type RegisteredRateLimiter struct {
-	rateLimiters map[unversioned.GroupVersion]struct{}
-	lock         sync.RWMutex
+	rateLimiters map[unversioned.GroupVersion]*sync.Once
 }
 
 // NewRegisteredRateLimiter returns a new RegisteredRateLimiater.
-func NewRegisteredRateLimiter() *RegisteredRateLimiter {
-	return &RegisteredRateLimiter{
-		rateLimiters: make(map[unversioned.GroupVersion]struct{}),
+// TODO: NewRegisteredRateLimiter is not dynamic. We need to find a better way
+// when GC dynamically change the resources it monitors.
+func NewRegisteredRateLimiter(resources []unversioned.GroupVersionResource) *RegisteredRateLimiter {
+	rateLimiters := make(map[unversioned.GroupVersion]*sync.Once)
+	for _, resource := range resources {
+		gv := resource.GroupVersion()
+		if _, found := rateLimiters[gv]; !found {
+			rateLimiters[gv] = &sync.Once{}
+		}
 	}
+	return &RegisteredRateLimiter{rateLimiters: rateLimiters}
 }
 
 func (r *RegisteredRateLimiter) registerIfNotPresent(gv unversioned.GroupVersion, client *dynamic.Client, prefix string) {
-	r.lock.RLock()
```

Comment:
I'm very confused about how this could cause lock contention unless it somehow always needed to register. At what line is it usually waiting for a lock? This line?